xformers/ops/fmha/tensor_with_seqlen.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+import itertools
+from typing import List, Optional, Sequence
+
+import torch
+
+
+def _copy_seqlen_info(
+    dst: "TensorWithSeqLen", src: "TensorWithSeqLen", deepcopy=False
+) -> None:
+    dst.max_seqlen = src.max_seqlen
+    if deepcopy:
+        dst.cu_seqlen = copy.deepcopy(src.cu_seqlen)
+        dst.cu_seqlen_py = copy.deepcopy(src.cu_seqlen_py)
+    else:
+        dst.cu_seqlen = src.cu_seqlen
+        dst.cu_seqlen_py = src.cu_seqlen_py
+        if dst.device != dst.cu_seqlen.device:
+            dst.cu_seqlen = dst.cu_seqlen.to(dst.device)
+
+
+def _find_arg_to_copy_metadata(cls, func, args, kwargs) -> Optional["TensorWithSeqLen"]:
+    first_cls = None
+    for a in itertools.chain(args, kwargs.values() if kwargs is not None else []):
+        if isinstance(a, cls):
+            if first_cls is not None:
+                assert first_cls.cu_seqlen is not None
+                assert first_cls.max_seqlen > 0
+                assert first_cls.cu_seqlen.shape == a.cu_seqlen.shape, f"Op: {func}"
+                assert first_cls.max_seqlen == a.max_seqlen, f"Op: {func}"
+            first_cls = a
+    assert first_cls is not None, f"Op: {func}"
+    return first_cls
+
+
+class TensorWithSeqLen(torch.Tensor):
+    max_seqlen: int
+    cu_seqlen: torch.Tensor
+    cu_seqlen_py: Sequence[int]
+
+    def __new__(cls, data):
+        t = torch.Tensor._make_subclass(cls, data)
+        t.cu_seqlen = None
+        t.cu_seqlen_py = []
+        t.max_seqlen = -1
+        t.extra_state = {
+            "last_func_called": "new",
+        }
+        return t
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        result = super().__torch_function__(func, types, args, kwargs)
+        if func.__name__ == "efficient_attention_forward_cutlass":
```

Comment:
nit: it might be more robust to compare the functions directly, as there is also a namespace that could play a role in here. But let's leave it like that for now